standards/application/dag-ack.md

```diff
@@ -0,0 +1,155 @@
+
+**Tip message**
+The latest message from a member. It can choose to reference or not reference the sender's previous message. It must reference the last seen messages from other `n` members in the group.
+
+**Control message**
+The message which is not input by user, but create and send automatically in application to facilitate our design.
+
+**Checkpoint message**
+It's a type of control message from a group member, which only contains references. Such message should not be persistent in any nodes, instead it's only used to find out if the referenced message is missing or not by other group members.
+
+**Request message**
+It's another type of control message for asking a message by its reference id. It can be persistent in store and sender node but not others. Group members will only fulfill the request message if store node doesn't have it. The sender node regularly check if pending request message gets fulfilled or not. To avoid the same request message flood by many members in one group, the sender node should not send the same request message if it already exists in the store node (using payload based query for such check).
+
+## Message Flows
+
+**Sending and retrieving messages**
+- Alice creates a group `G`
+- Bob, Charlie, Dave join `G`.
+- Alice sends message `a1` in `G` without any references.
+- Bob sends `b1` in `G` with reference to `a1`.
+- Charlies send `c1` with reference to `a1` and `b1`.
+- Dave sends `d1` with reference to `b1`, and `c1`.
+- Alice sends `a2` with reference to `a1`, `b1`, and `d1`.
+- Charlie receives `a2`, finds out that `d1` is not exist in local, he wants to retrieve the message from store node by the reference id.
+- If the `d1` did not exist in store node, charlie broadcasts a `request` message in the group.
+- Dave gets the `request` message and check if `d1` exists in local or not. 
+- Dave has `d1` and broadcast it to the network.
+
+
+## Payload Format
+
+### Tip Message
+
+The references for a message and checkpoint messages are generally encrypted in `WakuMessage.payload`. The first reference is always the previous message from the sender. The following references are the latest messages from other group members. The payload for `Tip Message` is defined as following,
+
+```python
+class Payload:
+    def __init__(self, references):
+        # omit other attributes
+        self.references = references
+```
+
+For test purpose, it can also be encoded in `WakuMessage.meta` for public analysis in following manner,
+
+```python
+concat(message_hash_1, message_hash_2, message_hash_3)
+```
+
+### Checkpoint Message
+
+The payload for `Checkpoint Message` is similar to Tip Message, but it only contains references, no other attributes.
+
+### Request Message
+
+The payload for `Request Message` is defiled as following,
```

Comment:
Just to be double sure I understand - say Alice believes she has sent messages `a1`..`a5`, each including a reference to the previous message. However, messages `a1`..`a4` actually failed to send due to a connectivity issue. The first successfully broadcast message is `a5`. Everyone in the group will see a reference to `a4`, which is missing, request this message and then recursively request all messages back in time to `a1`? Of course, we can try to avoid all recipients responding with a `request` at once by having them backoff randomly and check the store node in case another group member has requested the missing messages, but presumably there may still be a fairly large spike in activity for very large groups?